firefox-ios/WidgetKit/TopSites/TopSitesWidget.swift

```diff
@@ -21,78 +21,67 @@ struct TopSitesWidget: Widget {
+                    .fill(Color.clear)
+                    .overlay {
+                        image
+                            .resizable()
+                            .frame(width: imageSize, height: imageSize)
+                            .scaledToFit()
+                            .mask(maskShape)
+                    }
+                    .mask(maskShape)
+            } else {
+                emptySquare.frame(width: imageSize, height: imageSize)
             }
-        }
+        }.frame(width: itemSize, height: itemSize)
     }
 
     var maskShape: RoundedRectangle {
-        RoundedRectangle(cornerRadius: 5)
+        RoundedRectangle(cornerRadius: UX.maskShapeCornerRadius)
     }
 
     var emptySquare: some View {
         maskShape
-            .fill(UIColor(red: 0.85, green: 0.85, blue: 0.85, alpha: 0.3).color)
-            .frame(width: 60, height: 60)
-            .background(Color.clear).frame(maxWidth: .infinity)
+            .fill(UX.emptySquareFillColor)
+            .background(Color.clear)
     }
 
     var body: some View {
         VStack {
-            HStack {
-                if entry.sites.isEmpty {
-                    ForEach(0..<4, id: \.self) { _ in
-                        emptySquare
-                    }
-                } else if entry.sites.count > 3 {
-                    ForEach(entry.sites.prefix(4), id: \.url) { tab in
-                        topSitesItem(tab)
-                            .background(Color.clear).frame(maxWidth: .infinity)
-                    }
-                } else {
-                    ForEach(entry.sites[0...entry.sites.count - 1], id: \.url) { tab in
-                        topSitesItem(tab).frame(maxWidth: .infinity)
-                    }
-
-                    ForEach(0..<(4 - entry.sites.count), id: \.self) { _ in
-                        emptySquare
-                    }
-                }
-            }.padding([.top, .horizontal])
-            Spacer()
-            HStack {
-                if entry.sites.count > 7 {
-                    ForEach(entry.sites[4...7], id: \.url) { tab in
-                        topSitesItem(tab).frame(maxWidth: .infinity)
-                    }
-                } else {
-                    // Ensure there is at least a single site in the second row
-                    if entry.sites.count > 4 {
-                        ForEach(entry.sites[4...entry.sites.count - 1], id: \.url) { tab in
-                            topSitesItem(tab).frame(maxWidth: .infinity)
+            // Make a grid with 4 columns
+            GeometryReader { provider in
+                LazyVGrid(columns: (0..<4).map { _ in GridItem(.flexible(minimum: 0, maximum: .infinity)) },
+                          spacing: 0,
+                          content: {
+                    ForEach(0..<8) { index in
+                        if let site = entry.sites[safe: index] {
+                            topSitesItem(site, itemSize: provider.size.height / 2)
```

Comment:
You mentioned in the PR description that the images are now 30 x 30.

How can we guarantee this when we're using a relative size with the GeometryReader? It is not 100% clear what the image's final size is.

Line 74, itemSize = `provider.size.height / 2`
Then in topSitesItem, the imageSize  = `itemSize / 2`

But for emptySquare, its size is calculated at `provider.size.height / 4`.

If someone else looks at this code later, how can they quickly adjust the size to 50 x 50, for example? Maybe it would be better to pass in the exact size? Or if that would be difficult, can we do anything to make the size a little clearer, at least with the same scale between topSitesItem and emptySquare? 🤔 

P.S. I know the old code did not do this either 😅  It's just something Andy is asking the UX team about in terms of accessibility. So we may need to change the visual size later.

Reply:
Yes true, i forgot to mention that. Anyway i refactor the code so right now the size is more readable and understandable with the comments. The original size for a cell was `60 points` that is half of the minimum height for iOS device widget [https://developer.apple.com/design/human-interface-guidelines/widgets](widget sizes).
So that size was working for all devices, but for bigger ones the padding between the 2 rows were just increasing.

I Tried with setting a Standard size but that would be always depending on the smallest device. Instead of this i choose to set a standard scale of `1.0` that could be increased if we want to have bigger icons.
The base height for the icon is `1/4` of the widget height.
The scale also gets modified in function of the dynamic font size so it increase and decrease according the icon size.

i hope now it is better. But to provide an exact size for all devices is not possible, but we can say that the standard size with icon scale of `1.0` is `1/4` of the widget height.